open_spiel/games/crazy_eights.h

```diff
@@ -0,0 +1,227 @@
+  
+  private:
+    std::vector<Action> DealLegalActions() const;
+    std::vector<Action> PlayLegalActions() const;
+    void ApplyDealAction(int action);
+    void ApplyPlayAction(int action);
+    bool AfterPlayCard(int action);
+    void ScoreUp();
+
+    void Reshuffle();
+
+
+    std::vector<std::string> FormatHand(Player player) const;
+
+    std::string FormatAllHands() const;
+
+
+    Phase phase_ = Phase::kDeal;
+    int current_player_ = kInvalidPlayer;
+    int dealer_ = kInvalidPlayer;
+
+
+
+    // for the first card turned up, keep drawing if it is an eight
+    bool redraw_ = false;
+
+    // whether a player can pass
+    // it is true when (1) a player had already drawn max_draw_cards
+    // or (2) there is no card in the discard pile
+    bool can_pass_action_ = false;
+
+    // whether a player had already started to draw +2 cards
+    bool start_draw_twos_ = false;
+
+    // consecutive passes during a play
+    // if num_passes = num_player_ + 1, then the game ends
+    int num_passes_ = 0;
+
+
+    // the current accmulated +2 cards to be drawn
+    int num_draws_from_twos_left_ = 0;
+
+    // the number of consecutive draws for current_player_ so far
+    // this is not used for +2 cases
+    int num_draws_before_play_ = 0;
+
+    // the number of cards player can draw
+    int num_cards_left_;
+
+
+
+
+
+    int last_card_ = kInvalidAction;
+    int last_suit_ = -1;
+
+    bool nominate_suits_ = false;
+
+    int direction_ = 1;
+
+
+    bool reshuffle_;
+    int num_players_;
+    int max_draw_cards_;
+    int num_initial_cards_;
+    int num_decks_;
+    bool use_special_cards_;
+
+
+    std::vector<double> returns_;
+    std::array<int, kNumCards> dealer_deck_{};
+    std::vector<std::vector<int>> hands_;
+  
+};
+
+
+
+class CrazyEightsGame: public Game{
+  public:
+    explicit CrazyEightsGame(const GameParameters& params);
+    int NumDistinctActions() const override {return kDecideDealerActionBase + num_players_;}
+    int MaxChanceOutcomes() const override {return kDecideDealerActionBase + num_players_;}
+    std::unique_ptr<State> NewInitialState() const override {
+      return absl::make_unique<CrazyEightsState>(shared_from_this(), num_players_, max_draw_cards_, use_special_cards_, reshuffle_);
+    }
+    int NumPlayers() const override {return num_players_;}
+    double MinUtility() const override {return -504 * (num_players_ > 5? 2: 1);}
+    double MaxUtility() const override {return 0.0;}
+    std::vector<int> ObservationTensorShape() const override {
+      int num_decks = num_players_ > 5? 2: 1;
+      if(!use_special_cards_){
+        return {(num_decks + 1) * kNumCards + kNumCards + kNumSuits + (num_decks * kNumCards + 1) * (num_players_-1)};
+      } else return {(num_decks + 1) * kNumCards + kNumCards + kNumSuits + (num_decks * kNumCards + 1) * (num_players_-1) + 1};
+    
+    }
+    // In principle, the game can run indefinitely
+    int MaxGameLength() const override {return 10000;}
```

Comment:
I didn't see a terminating condition where if the game length hits the max it declares game over.

Reply:
Oh for this one, I think for crazy eights if we permit to reshuffle the deck after it get exhausted, then in principle the game could go indefinitely right? So for this case what should we put in MaxGameLength?

Reply:
In these cases we usually decide on a hard limit and end the game in a draw if it takes that long.